org.adempiere.base/src/org/compiere/model/MTable.java

```diff
@@ -1103,4 +1104,29 @@ public boolean hasCustomTree() {
 		return hasCustomTree.booleanValue();
 	}
 
+	/**
+	 * Get Partition Name of the table of the given level
+	 * @param tableName
+	 * @param primaryLevelOnly - if true, ignore the sub-partition, if exists
+	 * @return table partition name, or empty
+	 */
+	public static String getPartitionName(Properties ctx, String tableName, boolean primaryLevelOnly, String trxName) {
+		// collect partition key columns
+		String[] partitionColsAll = MTablePartition.getPartitionKeyColumns(ctx, tableName, trxName);
+		int level = primaryLevelOnly ? 1 : partitionColsAll.length;			
+		ArrayList<Object> partitionCols = new ArrayList<>();
+		for(int i = 0; i < level; i++) {
+			partitionCols.add(partitionColsAll[i]);
+		}
+		// parse partition name
+		StringBuilder partitionName = new StringBuilder();
+		partitionName.append(tableName);
+		for(Object partitionCol : partitionCols) {
+			String sPartitionCol = Objects.toString(partitionCol);
+			if(!Util.isEmpty(sPartitionCol)) {
+				partitionName.append("_").append(sPartitionCol);
+			}
+		}		

```

Comment:
as I've posted above, this is totally unnecessary long winded

Reply:
I see, modified the method according to your suggestion + added a null check for `tableName`, and a check for the length of `partitionColsAll` to avoid `ArrayIndexOutOfBoundsException` if `MTablePartition.getPartitionKeyColumns` returns an empty array